### ISSUE TITLE
nixos/syncthing: Can't declare devices in nixos config unless overrideDevice is false

### ISSUE THREAD

**Issue by Redlonghead** [NixOS/nixpkgs#326704]
### Describe the bug
When the option `services.syncthing.overrideDevices` is true you can not use `services.syncthing.settings.devices` to declare devices. 

### Steps To Reproduce
1. Set a device under `services.syncthing.settings.devices`
2. have the option `services.syncthing.overrideDevices = true`
3. rebuild config

### Expected behavior
Judging from the comment on the option I would be able to add the devices in a config permanently and add devices through the UI temporarily but the ones in the config would persist after a reboot. 

### Additional context
My config for syncthing
```nix
services.syncthing = {
  enable = true;
  user = "beacon";
  dataDir = "/home/beacon";
  configDir = "/home/beacon/.config/syncthing/";
  overrideDevices = false;
  overrideFolders = true;

  key = config.sops.secrets."syncthing/CB-FW/key".path;
  cert = config.sops.secrets."syncthing/CB-FW/cert".path;

  settings = {
    devices = { "CB-SNAS-01" = { id = "<ID>"; }; };
    folders = {
      "CB-SRC" = {
        path = "~/src";
        devices = [ "CB-SNAS-01" ];
        ignorePerms = false;
      };
      "CB-Documents" = {
        path = "~/Documents";
        devices = [ "CB-SNAS-01" ];
      };
    };
  };
};
```

### Notify maintainers
@Denperidge @johnhamelink 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.32, NixOS, 24.05 (Uakari), 24.05.20240531.63dacb4`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.2`
 - nixpkgs: `/nix/store/wzx1ba5hqqfa23vfrvqmfmkpj25p37mr-source`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by eclairevoyant** [comment]
Just to make sure, are those secrets listed above readable by the service user?

**Comment by Redlonghead** [comment]
Yes as I have those secrets owned by my user (beacon) and group of syncthing. I just followed the perms of the key and cert when they were first created so I have not had any problems so far.

**Comment by Denperidge** [comment]
Oh, I seemed to have missed the original notification, my apologies! I’ll try to take a look at how possible this is to implement

**Comment by Denperidge** [comment]
@Redhawk18 To double check: have you set overrideFolders to false? It is set to true by default, and would as such remove GUI-set things by default.
If you already did/the problem is elsewhere, can you post any relevant logs/output/screenshots/errors?

**Comment by Redhawk18** [comment]
> @Redhawk18 To double check: have you set overrideFolders to false? It is set to true by default, and would as such remove GUI-set shares by default. If you already did/the problem is elsewhere, can you post any relevant logs/output/screenshots/errors?

The aren't really any logs, the problem is the declarative options don't work on nixos for the client side. I believe I may be telling it to host two server folders? However what I linked made sense when I wrote it.

**Comment by Denperidge** [comment]
> > @Redhawk18 To double check: have you set overrideFolders to false? It is set to true by default, and would as such remove GUI-set shares by default. If you already did/the problem is elsewhere, can you post any relevant logs/output/screenshots/errors?
> 
> The aren't really any logs, the problem is the declarative options don't work on nixos for the client side. I believe I may be telling it to host two server folders? However what I linked made sense when I wrote it.
> 
> I turned off the override and did everything without nix and moved on.

I'll try to replicate the problem. Keep you posted!

**Comment by Redhawk18** [comment]
Thanks, I don't know if you can reproduce the problem with out two computers.

**Comment by AidanRB** [comment]
I'm having this issue as well. Two machines with basically identical configuration, here is one:

 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.37, NixOS, 24.11 (Vicuna), 24.11pre668759.8a3354191c0d`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.5`
 - channels(aidan): `"home-manager"`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

```nix
    syncthing = {
      enable = true;
      openDefaultPorts = true;
      settings = {
        options.relaysEnabled = false;
        folders.mariabackup = {
          path = "/backup/mariabackup";
          type = "sendonly";
          devices = [ "machine2" ];
        };

        devices.machine2 = {
          addresses = [ "tcp://address.for.machine2" ];
          id = "O73QLPW-machine2-id";
        };
      };
    };
```

In journalctl:
```
Connection from O73QLPW-machine2-id at address.for.machine1:22000 (tcp-server) rejected: unknown device
Established secure connection to O73QLPW at address.for.machine2:22000-address.for.machine1:22000/tcp-client/TLS1.3-...
Device O73QLPW client is "syncthing v1.27.9" named "" ...
Lost primary connection to O73QLPW at ... : reading length: EOF (0 remain)
Connection to O73QLPW at ... closed: reading length: EOF
Connection from O73QLPW-machine2-id ... rejected: unknown device
```

**Comment by dpc** [comment]
How I am debugging it:

```
> less /etc/systemd/system/syncthing-init.service                                                                                     
less /nix/store/xv1c234wdgliizk3pmxlfczzzbcqy8wh-merge-syncthing-config
```

pick the script, run manually with debugging on:

```
> env RUNTIME_DIRECTORY=/tmp bash -x /nix/store/xv1c234wdgliizk3pmxlfczzzbcqy8wh-merge-syncthing-config
```

Pick one of the curl commands:

```
+ curl -d '{"autoAcceptFolders":false,"deviceID":"...","id":"...","name":"..."}' -X POST 127.0.0.1:8384/rest/config/devices
```


Execute manually:

```
> ^curl -d '{"autoAcceptFolders":false,"deviceID":...","id":"...","name":"..."}' -X POST 127.0.0.1:8384/rest/config/devices
CSRF Error
```

**Comment by dpc** [comment]
Possibly related:

https://github.com/NixOS/nixpkgs/issues/260262

**Comment by dpc** [comment]
OK, now that I have more understanding how it works, it seems to me that once the cfsr error is triggered, the error status of curl command is successful anyway, which leads to some curl calls just randomly not working.

**Comment by dpc** [comment]
Hmmm... but it seems, it's like the api_key ... is just wrong. It is extracted from the config just fine, and it didn't change. Either it was not updated, or something else is wrong...

**Comment by dpc** [comment]
OK, so looking at my Syncthing UI in the browser, I can see the way the CSRF handling is completely different than `X-API-Key` that the script is trying to do. Not sure if it's alternative method, or it just no longer works, or never worked.

**Comment by dpc** [comment]
Yeah, so in the browser when you do `http://localhost:8384/` you'll get the cookies in the response, and then using them in other requests makes everything else work.

But from what I was able to tell, api key should work too... . And now I tried again and particular calls seems to be working, soooo WTH? :D . I still have some devices missing every time I ran that script.

**Comment by dpc** [comment]
Nvmd. I think I need to use `env RUNTIME_DIRECTORY=/home/dpc/.syncthing-data/.config/syncthing bash -x /nix/store/xv1c234wdgliizk3pmxlfczzzbcqy8wh-merge-syncthing-config`


After that I was able to debug more, and discover it was ... an user error, and I had somehow `0` an `O` changed in the id, and it was adding, then deleting the host as stale somehow. 

As far as I can tell 

```
overrideDevice=true
overrideFolder=true
```

works OK and the issue can be closed, so after making all that noise, I'll close it. Reopen if I missed something, or your think the issue is still there?
